### PR TITLE
Add citus_internal.distribute_object repair UDF

### PR DESCRIPTION
**Pull request by onurctirtir**
DESCRIPTION: Adds citus_internal.distribute_object() tool that can be used to manually propagate an object to workers.

This is a super-user only UDF that can be used by admins to
re-create an object as distributed in cluster-wide.

### CODE REVIEW

File: src/backend/distributed/commands/dependencies.c [hunk 1]
```diff
@@ -88,6 +156,190 @@ EnsureObjectAndDependenciesExistOnAllNodes(const ObjectAddress *target)
...
+ * out of the way before recreating it. The caller can pass forceRecreate to
+ * skip the per-node existence check and always (re)apply the DDL, which is
+ * useful to sync drift --e.g. role grants/options that were never propagated--
+ * onto nodes that already have the object.
+ *
+ * We deliberately accept the small TOCTOU window where the object might be
+ * created by someone else between our existence check and our own creation
+ * attempt: the worst case is that the create command fails loudly, which is
+ * acceptable for a superuser-only repair UDF.
+ *
```
Comment (onurctirtir) [inline src/backend/distributed/commands/dependencies.c]:
```suggestion
```

File: src/backend/distributed/commands/dependencies.c [hunk 2]
```diff
@@ -44,10 +49,73 @@ static int ObjectAddressComparator(const void *a, const void *b);
...
+ * By default a node is skipped if it already has the object; pass
+ * force_recreate => true to (re)apply the DDL on every node regardless.
+ */
+Datum
+citus_internal_distribute_object(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+	EnsureCoordinator();
+	EnsureSuperUser();
```
Comment (colm-mchugh) [inline src/backend/distributed/commands/dependencies.c:80]: Should `EnsureSuperUser();` precede `EnsureCoordinator()` ? As is, a non-superuser invoking the UDF on a worker gets error `operation is not allowed on this node` but getting `must be superuser` should take priority arguably.

File: src/backend/distributed/commands/dependencies.c [hunk 3]
```diff
@@ -44,10 +49,73 @@ static int ObjectAddressComparator(const void *a, const void *b);
...
+
+	Oid classId = PG_GETARG_OID(0);
+	Oid objectId = PG_GETARG_OID(1);
+	bool forceRecreate = PG_GETARG_BOOL(2);
+
+	ObjectAddress *objectAddress = palloc0(sizeof(ObjectAddress));
```
Comment (colm-mchugh) [inline src/backend/distributed/commands/dependencies.c]: Nit: use the stack instead 
```
ObjectAddress objectAddress = { 0 } ;
```

File: src/backend/distributed/commands/dependencies.c [hunk 4]
```diff
@@ -44,10 +49,73 @@ static int ObjectAddressComparator(const void *a, const void *b);
...
+ */
+Datum
+citus_internal_distribute_object(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+	EnsureCoordinator();
+	EnsureSuperUser();
+
+	Oid classId = PG_GETARG_OID(0);
+	Oid objectId = PG_GETARG_OID(1);
```
Comment (colm-mchugh) [inline src/backend/distributed/commands/dependencies.c:107]: Anything to be said for also passing in an `objectSubId` ? I've no particular reason other than completeness for this suggestion, but including it with a default value of 0 makes the treatment of `objectSubId` explicit, and future-proofs the API.

File: src/backend/distributed/commands/dependencies.c [hunk 5]
```diff
@@ -88,6 +156,175 @@ EnsureObjectAndDependenciesExistOnAllNodes(const ObjectAddress *target)
...
+		if (forceRecreate ||
+			!RemoteCommandOnNodeReturnsRow(nodeName, nodePort, existenceCheckCommand))
+		{
+			SendCommandListToWorkerOutsideTransaction(nodeName, nodePort,
+													  CitusExtensionOwnerName(),
+													  ddlCommands);
+		}
+	}
+
+	MarkObjectDistributedViaSuperUser(target);
```
Comment (colm-mchugh) [inline src/backend/distributed/commands/dependencies.c:252]: This function (`EnsureObjectExistsOnAllNodes()`) commits per-node DDL via `SendCommandListToWorkerOutsideTransaction()` but the subsequent `MarkObjectDistributedViaSuperUser(target)` here runs inside the caller's coordinated transaction. Iis it possible that the caller's transaction could abort after one or more workers already committed the DDL?  E.g. a Ctrl+C between the DDL loop and the metadata write, or a coordinator crash before 2PC prepare, or an operator-issued ROLLBACK in the same explicit transaction. If so, then we land in the "workers have the object, no pg_dist_object row anywhere" state. A concurrent `citus_add_node` in that window will not replay the object onto the new node (metadata-sync uses `pg_dist_object`), silently reintroducing the drift this UDF was designed to repair. Retry does eventually reconverge (probe returns "row present" → DDL skipped → mark succeeds), but the intermediate window is unbounded. Is this a viable concern ? If so how about requiring `PreventInTransactionBlock(true, "citus_internal.distribute_object")` at the top of the UDF so an operator-issued ROLLBACK cannot leave the split state?

File: src/backend/distributed/commands/dependencies.c [hunk 6]
```diff
@@ -88,6 +156,175 @@ EnsureObjectAndDependenciesExistOnAllNodes(const ObjectAddress *target)
...
+	foreach_declared_ptr(objectArg, objectArgs)
+	{
+		appendStringInfo(command, "%s%s", firstArg ? "" : ", ",
+						 quote_literal_cstr(objectArg));
+		firstArg = false;
+	}
+
+	appendStringInfoString(command, "]::text[])");
+
+	return command->data;
```
Comment (colm-mchugh) [inline src/backend/distributed/commands/dependencies.c:304]: The `SELECT 1 ...` probe treats any successful name resolution on the worker as "object present", is it sufficient to match on the name only ? Can we include a `NOTICE` naming the node so clients can see which nodes were skipped.. Also a test suggestion for `citus_internal_distribute_object.sql`:
- Use `master_run_on_worker` to create a same-named function with a different body on one worker
- call `distribute_object` without `force_recreate` and assert the worker's body remains drifted
- calls with `force_recreate = true` and asserts no drift

File: src/backend/distributed/commands/dependencies.c [hunk 7]
```diff
@@ -88,6 +156,175 @@ EnsureObjectAndDependenciesExistOnAllNodes(const ObjectAddress *target)
...
+
+	/*
+	 * Otherwise the command raised a statement-level error, which we treat as
+	 * "no rows returned" so that an existence probe reports the object as absent.
+	 */
+	PQclear(result);
+	ForgetResults(connection);
+
+	return returnedRow;
+}
```
Comment (colm-mchugh) [inline src/backend/distributed/commands/dependencies.c:362]: This function (`RemoteCommandOnNodeReturnsRow()`) opens the connection with `OUTSIDE_TRANSACTION`, does not add `FORCE_NEW_CONNECTION` and does not claim the connection exclusively and does not `ROLLBACK` the connection after a statement-level error. 
Concerned that this is insufficient given the warning in `connection_management.h`: _a connection returned with OUTSIDE_TRANSACTION is not guaranteed to stay outside of transactions … callers should also mark the connection exclusive by using ClaimConnectionExclusively()_  Consequently, `GetNodeUserDatabaseConnection()` could hand us a cached connection that is part of the caller's coordinated transaction. A statement error could then put the coordinated remote transaction into error, contradicting the _never rolls back ..._  claim in the function comment. Why not follow the advice of `connection_managment.h` and change the connection set-up to:
```
int connectionFlags = OUTSIDE_TRANSACTION | FORCE_NEW_CONNECTION;
MultiConnection *connection =
    GetNodeUserDatabaseConnection(connectionFlags, nodeName, nodePort,
                                  CitusExtensionOwnerName(), NULL);
ClaimConnectionExclusively(connection);
```
and then throw in an `UnclaimConnection(connection);` along with the `PQclear()` and `ForgetResults()`

File: src/backend/distributed/sql/udfs/citus_internal_distribute_object/latest.sql [hunk 8]
```diff
@@ -0,0 +1,7 @@
+CREATE OR REPLACE FUNCTION citus_internal.distribute_object(classid oid, objid oid, force_recreate boolean DEFAULT false)
+    RETURNS void
+    LANGUAGE C STRICT
+    AS 'MODULE_PATHNAME', $$citus_internal_distribute_object$$;
+COMMENT ON FUNCTION citus_internal.distribute_object(oid, oid, boolean)
+    IS 'recreate the given object on all worker nodes and record it in pg_dist_object on all nodes';
```
Comment (colm-mchugh) [inline src/backend/distributed/sql/udfs/citus_internal_distribute_object/latest.sql]: Can we enhance the comment to note that dependencies are not propagated and that the DDL and metadata write are not atomic ?

File: src/test/regress/sql/citus_internal_distribute_object.sql [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+RESET citus.enable_metadata_sync;
+SELECT bool_and(success) AS srv_iv_placed FROM master_run_on_worker(ARRAY['localhost']::text[], ARRAY[:one_worker_port]::int[], ARRAY[$$SET citus.enable_ddl_propagation TO off; CREATE SERVER distobj_srv4 FOREIGN DATA WRAPPER postgres_fdw OPTIONS (host 'localhost')$$]::text[], false);
+SELECT bool_and(result::int = 1) AS srv_iv_partial FROM run_command_on_workers($$SELECT count(*) FROM pg_foreign_server WHERE srvname = 'distobj_srv4'$$);
+SELECT citus_internal.distribute_object('pg_foreign_server'::regclass::oid, oid) FROM pg_foreign_server WHERE srvname = 'distobj_srv4';
+SELECT bool_and(result::int = 1) AS srv_iv_full FROM run_command_on_workers($$SELECT count(*) FROM pg_foreign_server WHERE srvname = 'distobj_srv4'$$);
+DROP SERVER distobj_srv4;
+DROP SERVER distobj_srv;
+DROP EXTENSION postgres_fdw;
+DROP ROLE distobj_role;
+DROP PUBLICATION distobj_pub;
```
Comment (colm-mchugh) [inline src/test/regress/sql/citus_internal_distribute_object.sql:333]: Additional test suggestions, aimed at negative-path coverage:

- Call on an unsupported catalog OID gets `SupportedDependencyByCitus` error.
- Call on an OCLASS with an empty DDL list (e.g. a bare `OCLASS_CONSTRAINT`) gets empty DDL error
- A sequence such as `SET ROLE nonsuper; SELECT citus_internal.distribute_object(...); `gets `EnsureSuperUser` error
- Attempting `SELECT run_command_on_workers($$ SELECT citus_internal.distribute_object(...) $$);` get `EnsureCoordinator` on each worker
- A non superuser without `EXECUTE` calling the UDF hits the `REVOKE` gate

File: src/backend/distributed/commands/dependencies.c [hunk 10]
```diff
@@ -44,10 +49,73 @@ static int ObjectAddressComparator(const void *a, const void *b);
...
+citus_internal_distribute_object(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+	EnsureCoordinator();
+	EnsureSuperUser();
+
+	Oid classId = PG_GETARG_OID(0);
+	Oid objectId = PG_GETARG_OID(1);
+	bool forceRecreate = PG_GETARG_BOOL(2);
+
```
Comment (colm-mchugh) [inline src/backend/distributed/commands/dependencies.c:110]: Should we check for and only proceed if `citus.enable_metadata_sync` is on ?

File: src/backend/distributed/commands/dependencies.c (hunk 4 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/dependencies.c:107]: yes, this is for completeness, let me make it default to 0

File: src/backend/distributed/commands/dependencies.c (hunk 2 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/dependencies.c:80]: Good catch — reordered so `EnsureSuperUser()` runs before `EnsureCoordinator()`. A non-superuser now gets the superuser error even when the UDF is invoked on a worker node, instead of the coordinator-only error. Added a negative test that runs the UDF on a worker to lock the ordering in.

File: src/backend/distributed/commands/dependencies.c (hunk 3 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/dependencies.c]: Done — switched to a stack `ObjectAddress` initialized via `ObjectAddressSubSet` and pass `&objectAddress` at the call sites, no more `palloc0`.

File: src/backend/distributed/commands/dependencies.c (hunk 4 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/dependencies.c:107]: Added `objsubid int DEFAULT 0` — the signature is now `distribute_object(classid oid, objid oid, objsubid int DEFAULT 0, force_recreate boolean DEFAULT false)`. The 2-arg call form still works via the defaults, and the C side reads it into the `ObjectAddress` subid.

File: src/backend/distributed/commands/dependencies.c (hunk 5 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/dependencies.c:252]: Added `PreventInTransactionBlock(true, ...)`. This closes the operator-`ROLLBACK` window you flagged. One thing worth noting: it doesn't close the crash-*between* the worker DDL and `MarkObjectDistributed` — but that case is self-healing, since re-running the repair UDF reconverges (workers already have the object so it's skipped, then pg_dist_object gets marked). That felt acceptable for a superuser repair tool, but happy to add a stronger guarantee if you'd prefer.

File: src/backend/distributed/commands/dependencies.c (hunk 6 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/dependencies.c:304]: The by-name probe is intentional: OIDs differ across nodes, so the name is the only stable identifier we can match on. `force_recreate` is the escape hatch for the case the probe can't see — a same-named object whose body has drifted. To cover it I added (a) a per-skipped-node NOTICE, and (b) a test that drifts one worker's function body, asserts it survives a no-force call, then asserts `force_recreate := true` re-syncs every worker.

File: src/backend/distributed/commands/dependencies.c (hunk 7 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/dependencies.c:362]: Agreed — kept `OUTSIDE_TRANSACTION` and added `FORCE_NEW_CONNECTION` on top, plus `ClaimConnectionExclusively`/`UnclaimConnection`. So the probe can never reuse a cached coordinated-transaction connection, and a statement error on the probe can't poison the caller's remote transaction state.

File: src/backend/distributed/sql/udfs/citus_internal_distribute_object/latest.sql (hunk 8 above)
Comment (onurctirtir) [reply src/backend/distributed/sql/udfs/citus_internal_distribute_object/latest.sql]: Expanded the COMMENT to spell out that dependencies are not propagated and that the worker DDL and the pg_dist_object writes are not performed atomically.

File: src/test/regress/sql/citus_internal_distribute_object.sql (hunk 9 above)
Comment (onurctirtir) [reply src/test/regress/sql/citus_internal_distribute_object.sql:333]: Added negative-path coverage: (a) an unsupported type (materialized view) hits the `SupportedDependencyByCitus` error; (b) a supported-but-empty-DDL object (a plain, non-distributed table) hits "could not generate DDL commands"; (c) `SET ROLE` to a non-superuser with EXECUTE granted still hits the superuser check; (d) a non-superuser without EXECUTE hits the REVOKE gate; (e) running it on a worker hits the coordinator-only check. I also changed the empty-DDL error to print the object identity instead of a raw OID so the expected output is stable across runs.

File: src/backend/distributed/commands/dependencies.c (hunk 10 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/dependencies.c:110]: Added a hard error when `citus.enable_metadata_sync` is off. Otherwise `MarkObjectDistributedViaSuperUser` silently skips worker propagation and we'd leave the cluster half-repaired — object recreated on workers, but pg_dist_object only marked on the coordinator.

File: src/backend/distributed/commands/dependencies.c [hunk 11]
```diff
@@ -44,10 +50,96 @@ static int ObjectAddressComparator(const void *a, const void *b);
...
+	/*
+	 * The pg_dist_object records are only propagated to the other nodes when
+	 * metadata syncing is enabled. Without it we would create the object on the
+	 * workers but record it only on the coordinator, so we refuse to run rather
+	 * than leave the cluster in that half-repaired state.
+	 */
+	if (!EnableMetadataSync)
+	{
+		ereport(ERROR, (errmsg("citus_internal.distribute_object requires "
+							   "citus.enable_metadata_sync to be enabled")));
```
Comment (colm-mchugh) [inline src/backend/distributed/commands/dependencies.c]: An `errhint()` telling the user to `SET citus.enable_metadata_sync To on` could be helpful, but it may be obvious what to do from the error message, up to you whether to add or not.

**Review by colm-mchugh — APPROVED** [review]
Lgtm, thx for addressing CR feedback.

File: src/backend/distributed/commands/dependencies.c (hunk 11 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/dependencies.c]: Good idea — added an `errhint()` telling the user to `SET citus.enable_metadata_sync TO on` before calling the function. I also noticed this gate had no test coverage, so I added a negative test that disables metadata sync, calls the UDF, and asserts both the error and the new hint. Done in a926a3469.
